### PR TITLE
move inter pod affinity predicate logic to its Filter plugin

### LINKED ISSUE #86460: Move Pod Affinity predicate logic to its Filter plugin [Migration Phase 2]
/assign @ahg-g 
/sig scheduling
/priority important-soon

This includes moving the metadata logic into PreFilter.

Part of #85822

### PR DESCRIPTION
**Pull request by ahg-g**
**What type of PR is this?**

/kind cleanup

**Which issue(s) this PR fixes**:
Fixes #86460

```release-note
NONE
```

/cc @Huang-Wei @alculquicondor

### CODE REVIEW

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go [hunk 1]
```diff
@@ -55,39 +59,303 @@ var _ framework.FilterPlugin = &InterPodAffinity{}
...
+}
+
+// TODO(Huang-Wei): It might be possible to use "make(map[topologyPair]*int64)" so that
+// we can do atomic additions instead of using a global mutext, however we need to consider
+// how to init each topologyToMatchedTermCount.
+type topologyPair struct {
+	key   string
+	value string
+}
+type topologyToMatchedTermCount map[topologyPair]int64
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: I don't like how we use `map[string]map[string]int64` for scoring, but `map[topologyPair]int64` for filtering. We should probably just pick one (the one that gives us the best performance). But that's for a follow up.

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go [hunk 2]
```diff
@@ -55,39 +59,303 @@ var _ framework.FilterPlugin = &InterPodAffinity{}
...
 var _ framework.ScorePlugin = &InterPodAffinity{}
 
-const (
-	// Name is the name of the plugin used in the plugin registry and configurations.
-	Name = "InterPodAffinity"
-
-	// preFilterStateKey is the key in CycleState to InterPodAffinity pre-computed data for Filtering.
-	// Using the name of the plugin will likely help us avoid collisions with other plugins.
-	preFilterStateKey = "PreFilter" + Name
+// InterPodAffinity is a plugin that checks inter pod affinity
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: now I feel like this file is too big. But the plugin lives in a folder, so we can do something like:

```
interpodaffinity/
  plugin.go
  filtering.go
  scoring.go
```

WDYT?

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go [hunk 3]
```diff
@@ -55,39 +59,303 @@ var _ framework.FilterPlugin = &InterPodAffinity{}
...
+// A processed version of v1.PodAffinityTerm.
+type affinityTerm struct {
+	namespaces  sets.String
+	selector    labels.Selector
+	topologyKey string
+}
+
+// getAffinityTermProperties receives a Pod and affinity terms and returns the namespaces and
+// selectors of the terms.
+func getAffinityTermProperties(pod *v1.Pod, terms []v1.PodAffinityTerm) (properties []*affinityTerm, err error) {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: s/properties/terms

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: sg, I was thinking the same thing.

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: yeah, priority and predicate evolved separately, make sense to unify them, let me see if I can do it in this PR.

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go [hunk 4]
```diff
@@ -146,25 +426,258 @@ func getPodAffinityMetadata(cycleState *framework.CycleState) (*predicates.PodAf
...
+	return nil
 }
 
 // A "processed" representation of v1.WeightedAffinityTerm.
 type weightedAffinityTerm struct {
-	namespaces  sets.String
-	selector    labels.Selector
-	weight      int32
-	topologyKey string
+	affinityTerm
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: +1
now this looks just like v1 :D

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go [hunk 5]
```diff
@@ -55,39 +59,303 @@ var _ framework.FilterPlugin = &InterPodAffinity{}
...
+
+// getAffinityTermProperties receives a Pod and affinity terms and returns the namespaces and
+// selectors of the terms.
+func getAffinityTermProperties(pod *v1.Pod, terms []v1.PodAffinityTerm) (properties []*affinityTerm, err error) {
+	if terms == nil {
+		return properties, nil
+	}
+
+	for _, term := range terms {
+		namespaces := priorityutil.GetNamespacesFromPodAffinityTerm(pod, &term)
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: it looks like this block can be factored out and shared with postfilter?

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go (hunk 3 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: done.

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go (hunk 2 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: done.

File: pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go (hunk 1 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/interpodaffinity/interpod_affinity.go]: lets do that as a follow up

**Comment by alculquicondor** [comment]
great. Let's just wait for @Huang-Wei comments.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+var _ framework.PreFilterPlugin = &InterPodAffinity{}
+var _ framework.FilterPlugin = &InterPodAffinity{}
+var _ framework.PostFilterPlugin = &InterPodAffinity{}
+var _ framework.ScorePlugin = &InterPodAffinity{}
+
+// InterPodAffinity is a plugin that checks inter pod affinity
+type InterPodAffinity struct {
+	sharedLister          schedulerlisters.SharedLister
+	hardPodAffinityWeight int32
+	sync.Mutex
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: The mutex is never used. We defined local Mutexs in `getTPMapMatchingExistingAntiAffinity` and `getTPMapMatchingIncomingAffinityAntiAffinity`.

**Review by Huang-Wei — COMMENTED** [review]
LGTM generally. Just one nit.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go (hunk 6 above)
Comment (ahg-g) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: it is actually used by the score logic.

Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: oh, true... Ignore my comment then.

**Pull request merged**